src/Compilers/CSharp/Portable/Binder/DecisionDagBuilder.cs

```diff
@@ -825,14 +836,24 @@ DagState uniqifyState(ImmutableArray<StateForCase> cases, ImmutableDictionary<Bo
                     // Select the next test to do at this state, and compute successor states
                     switch (state.SelectedTest = state.ComputeSelectedTest())
                     {
+                        case BoundDagAssignmentEvaluation e
+                            when state.RemainingValues.TryGetValue(e.Input, out IValueSet? currentValues) |
+                                 state.RemainingValues.TryGetValue(e.Target, out IValueSet? targetValues):
+                            IValueSet commonValues = (currentValues, targetValues) is (IValueSet v1, IValueSet v2)
+                                ? v1.Intersect(v2) : currentValues ?? targetValues ?? throw ExceptionUtilities.Unreachable;
+                            var builder = ArrayBuilder<KeyValuePair<BoundDagTemp, IValueSet>>.GetInstance(2);
+                            builder.Add(new KeyValuePair<BoundDagTemp, IValueSet>(e.Input, commonValues));
```

Comment:
nit: In an assignment we would not update the values of the input. May need to come up with a better name for `BoundDagAssignmentEvaluation` type. Maybe `BoundMergeDagTemps`?

Reply:
I can remove this line and not affect any tests. Is that expected?

Reply:
>In an assignment we would not update the values of the input.

That's right. 

The scenario I had in mind was that if T0 and T1 were related, we want the values from T1 to be considered for a pair of T0 and T2, however, T1 and T2 will be related on their own turn if they had to, so we don't really need to update those values here.

Reply:
BoundMergeDagTemps wouldn't make sense in the context of lowering. I'd stick with with the current name for now if you don't think it's too confusing.